go/vt/vtctl/grpcvtctldserver/server_test.go

```diff
@@ -8811,3 +8813,633 @@ func TestValidateShard(t *testing.T) {
+				Cells:    []string{"zone1"},
+				VSchema:  tt.vschema,
+			}
+			_, err = vtctld.ApplyVSchema(ctx, req3)
+			require.NoError(t, err, "error ApplyingVSchema")
+
+			req4 := &vtctldata.ValidateSchemaKeyspaceRequest{
+				Keyspace:       tt.keyspace,
+				SkipNoPrimary:  false,
+				ExcludeTables:  nil,
+				IncludeVSchema: tt.includeVSchema,
+				IncludeViews:   true,
+			}
+			_, err = vtctld.ValidateSchemaKeyspace(ctx, req4)
+			if tt.shouldErr {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err, "error in ValidateSchemaKeyspace")
+		})
+	}
+}
+
+func TestValidateSchemaShard(t *testing.T) {
+	ctx := context.Background()
+
+	tests := []struct {
+		name           string
+		schemas        []*tabletmanagerdatapb.SchemaDefinition
+		vschema        *vschemapb.Keyspace
+		includeVSchema bool
+		shouldErr      bool
+	}{
+		{
+			name: "should pass",
+			schemas: []*tabletmanagerdatapb.SchemaDefinition{
+				{
+					TableDefinitions: []*tabletmanagerdatapb.TableDefinition{
+						{
+							Name:    "t1",
+							Columns: []string{"c1"},
+						},
+					},
+				},
+			},
+			vschema: &vschemapb.Keyspace{
+				Sharded: true,
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						Columns: []*vschemapb.Column{
+							{
+								Name: "c1",
+								Type: querypb.Type_BINARY,
+							},
+						},
+						ColumnVindexes: []*vschemapb.ColumnVindex{
+							{
+								Column: "c1",
+								Name:   "c1",
+							},
+						},
+					},
+				},
+				Vindexes: map[string]*vschemapb.Vindex{
+					"c1": {
+						Type: "hash",
+					},
+				},
+			},
+			includeVSchema: false,
+			shouldErr:      false,
+		},
+		{
+			name: "different schemas",
+			schemas: []*tabletmanagerdatapb.SchemaDefinition{
+				{
+					TableDefinitions: []*tabletmanagerdatapb.TableDefinition{
+						{
+							Name:    "t1",
+							Columns: []string{"c1"},
+						},
+					},
+				},
+				{
+					TableDefinitions: []*tabletmanagerdatapb.TableDefinition{
+						{
+							Name:    "t2",
+							Columns: []string{"c1"},
+						},
+					},
+				},
+				{
+					TableDefinitions: []*tabletmanagerdatapb.TableDefinition{
+						{
+							Name:    "t3",
+							Columns: []string{"c1"},
+						},
+					},
+				},
+			},
+			vschema: &vschemapb.Keyspace{
+				Sharded: true,
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						Columns: []*vschemapb.Column{
+							{
+								Name: "c1",
+								Type: querypb.Type_BINARY,
+							},
+						},
+						ColumnVindexes: []*vschemapb.ColumnVindex{
+							{
+								Column: "c1",
+								Name:   "c1",
+							},
+						},
+					},
+				},
+				Vindexes: map[string]*vschemapb.Vindex{
+					"c1": {
+						Type: "hash",
+					},
+				},
+			},
+			includeVSchema: false,
+			shouldErr:      true,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Test Setup
+			ts, _ := memorytopo.NewServerAndFactory("zone1")
+			tmc := &testutil.TabletManagerClient{
```

Comment:
you probably want this in the `tests` slice, so you can more easily configure individual test cases (same for the other functions you're testing)